### PR TITLE
[REVIEW] Port search.pxy to libcudf++ and support multi-column searchsorted

### LINKED ISSUE #3943: [FEA] Port search.pyx to use non-legacy libcudf APIs

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #3943 

- Ports cython wrappers to libcudf++
- Updates Python APIs (provides support for multi-column, DataFrame-level, `searchsorted`)

**TODO**:
- [x] Add multi-column/`DataFrame` tests for `searchsorted`
- [ ] Finish porting `contains`

### CODE REVIEW

File: python/cudf/cudf/core/frame.py [hunk 1]
```diff
@@ -219,6 +220,41 @@ def _unaryop(self, op):
...
+
+        Returns
+        -------
+        1-D cupy array of insertion points
+        """
+        # Convert column-typed `values` to Series
+        if isinstance(values, ColumnBase):
+            from cudf.core.series import Series
+
+            values = Series(values)
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: We shouldn't need to handle a Column of values in this function. If Column is being passed into here it should call `.as_frame()` first

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -778,6 +778,15 @@ def __cuda_array_interface__(self):
 
         return output
 
+    def searchsorted(
+        self, value, side="left", ascending=True, na_position="last"
+    ):
+        values = cudf.Series(as_column(value))
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py]:
```suggestion
        values = as_column(value).as_frame()
```

File: python/cudf/cudf/core/column/column.py [hunk 3]
```diff
@@ -778,6 +778,15 @@ def __cuda_array_interface__(self):
...
+        col = cudf.Series(self).searchsorted(
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py]:
```suggestion
        col = self.as_frame().searchsorted(
```

File: python/cudf/cudf/_libxx/includes/search.pxd [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from cudf._libxx.lib cimport *
+
+cdef extern from "cudf/search.hpp" namespace "cudf::experimental" nogil:
+
+    cdef unique_ptr[column] lower_bound(
+        table_view t,
+        table_view values,
+        vector[order] column_order,
+        vector[null_order] null_precedence,
+    )
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf/cudf/_libxx/includes/search.pxd]: @kkraus14 - Any idea why cythonize fails for me when I add `except +` here?

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/_libxx/includes/search.pxd]: @rjzamora do you have a stack trace? I'm guessing a missing `move` operator which is what we've typically seen.

File: python/cudf/cudf/_libxx/search.pyx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    c_order = order.ASCENDING if ascending else order.DESCENDING
+    c_null_order = (
+        null_order.AFTER if na_position=="last" else null_order.BEFORE
+    )
+    c_column_order = vector[order](table._num_columns, c_order)
+    c_null_precedence = vector[null_order](table._num_columns, c_null_order)
+
+    if side == 'left':
+        with nogil:
+            c_result = (
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/search.pyx]: Likely need the moves here to resolve the `except +` issues you're seeing:
```suggestion
            c_result = move(
```

Comment (rjzamora) [reply python/cudf/cudf/_libxx/search.pyx]: Perfect - Thanks!

File: python/cudf/cudf/core/column/numerical.py [hunk 6]
```diff
@@ -53,7 +54,10 @@ def __contains__(self, item):
                 return False
         except Exception:
             return False
-        return libcudf.search.contains(self, item)
+        # TODO: Use `scalar`-based `contains` wrapper
+        return libcudfxx.search.contains(
+            self, column.as_column([item], dtype=self.dtype)
+        ).any()
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py]: cc @shwina since he's currently looking at handling scalar

**Comment by rjzamora** [comment]
> @rjzamora this looks really clean, are you adding anything else to it / need any more changes?

I'm not planning anything else besides the switch to the `scalar`-based `contains` wrapper when possible.  At the moment, this should cover the port of existing `searchsorted` and `contains` functionality (and also adds dataframe-level `searchsorted` support).

**Comment by kkraus14** [comment]
> > @rjzamora this looks really clean, are you adding anything else to it / need any more changes?
> 
> I'm not planning anything else besides the switch to the `scalar`-based `contains` wrapper when possible. At the moment, this should cover the port of existing `searchsorted` and `contains` functionality (and also adds dataframe-level `searchsorted` support).

We can handle the `scalar` change in a follow up, approving this as this looks good to go now. Thanks for tackling this!

**Pull request merged**